hibernate-core/src/main/java/org/hibernate/query/sqm/internal/TypecheckUtil.java

```diff
@@ -508,6 +508,15 @@ else if ( isNumberArray( leftNodeType ) ) {
 		}
 	}
 
+	public static boolean isGeneric(SqmExpression<?> arg) {
+		return arg instanceof SqmPath && ( (SqmPath) arg ).getModel() instanceof SqmPathSource
+				&& ( (SqmPathSource) ( (SqmPath) arg ).getModel() ).isGeneric();
+	}
+
+	public static SqmExpressible<?> getExpressibleNodeType(SqmExpression<?> arg) {
+		return isGeneric( arg ) ? ( (SqmPath) arg ).getResolvedModel() : arg.getNodeType();
+	}
+
```

Comment:
We should not need to do all this `instanceof` checks and typecasts, as the SQM tree objects should already expose the concrete-generic type of attribute paths. Please try using `SqmTypedNode#getExpressible` instead of this helper method wherever needed.

Reply:
Now I've removed `getExpressibleNodeType` helper method. In case when `isGeneric` is returning false, current handling is left as is. Only in cases when `getExpressibleNodeType` return true `SqmTypedNode#getExpressible` is used. Otherwise, some current tests can fail.

Reply:
That should not be the case, and the `isGeneric()` helper should also be removed. Always using `getExpressible()` and relying on the implementations to resolve the concrete-generic type is the correct path forward, though there might be some other bugs or missing overrides of that method that are causing the other tests to fail. Could you please try and look into them?

Reply:
OK, will try to look into those problems, but I am not sure WHAT RESULT WILL BE.

Reply:
@mbladel I've noticed that `org.hibernate.query.sqm.SqmExpressible#getRelationalJavaType` defaulted to `getExpressibleJavaType` and implemented only in `org.hibernate.metamodel.model.domain.internal.BasicSqmPathSource`, while `getExpressionableJavaType` is often implemented as delegate to something. Is this correct?

If I implement `getRelationalJavaType` as delegate in `org.hibernate.metamodel.model.domain.internal.SingularAttributeImpl` and `org.hibernate.query.sqm.tree.domain.SqmBasicValuedSimplePath`, existing tests will again pass, and only failing tests will be newly addres that are selecting functions. However, I am not sure that this is correct approach, so I am not commiting anything.

Reply:
I do believe that is correct, indeed. We also noticed this recently in https://github.com/hibernate/hibernate-orm/pull/8594 and made similar changes. We could wait for that PR to be merged to go ahead with this one, or you can cherry-pick the changes locally, but the fact that it solves this too increases my confidence that the overrides are correct.

> only failing tests will be newly addres that are selecting functions

It would be great if you could also look into those - I believe the solution might be very similar to what we did here, i.e. implying `SqmTypedNode#getExpressible` for functions that inherit the return type from their arguments as well.

Reply:
Those changes are exactly what I've used to fix some of tests. I will not make any changes in classes changed in 8594.

Now only problem is with function argument. I have no idea how to get proper type of overridden generic argument without checking if it is generic. Method that is causing problems is `org.hibernate.query.sqm.produce.function.StandardFunctionReturnTypeResolvers#getArgumentExpressible`

Reply:
You could try using `getExpressible()` in that function instead of `getNodeType()`, just like I suggested for the other call sites, as the first method should correctly resolve to the "concrete" type.

Reply:
Right, `getExpressible()`, and then `getSqmType()` when not null. Previously, I've tried former, but missed later.
Now all tests will pass.